src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -5182,41 +5190,40 @@ public void decommission(boolean force) throws InterruptedException
             }
 
             startLeaving();
-            long timeout = Math.max(RING_DELAY_MILLIS, BatchlogManager.instance.getBatchlogTimeout());
+            long timeout = Math.max(RING_DELAY_MILLIS, BatchlogManager.getBatchlogTimeout());
             setMode(Mode.LEAVING, "sleeping " + timeout + " ms for batch processing and pending range setup", true);
-            Thread.sleep(timeout);
+            Uninterruptibles.sleepUninterruptibly(timeout, MILLISECONDS);
+
+            unbootstrap();
 
-            Runnable finishLeaving = new Runnable()
+            // shutdown cql, gossip, messaging, Stage and set state to DECOMMISSIONED
+
+            shutdownClientServers();
+            Gossiper.instance.stop();
+            try
             {
-                public void run()
-                {
-                    shutdownClientServers();
-                    Gossiper.instance.stop();
-                    try
-                    {
-                        MessagingService.instance().shutdown();
-                    }
-                    catch (IOError ioe)
-                    {
-                        logger.info("failed to shutdown message service: {}", ioe);
-                    }
+                MessagingService.instance().shutdown();
+            }
+            catch (IOError ioe)
+            {
+                logger.info("failed to shutdown message service", ioe);
+            }
 
-                    Stage.shutdownNow();
-                    SystemKeyspace.setBootstrapState(SystemKeyspace.BootstrapState.DECOMMISSIONED);
-                    setMode(Mode.DECOMMISSIONED, true);
-                    // let op be responsible for killing the process
-                }
-            };
-            unbootstrap(finishLeaving);
-        }
-        catch (InterruptedException e)
-        {
-            throw new UncheckedInterruptedException(e);
+            Stage.shutdownNow();
+            SystemKeyspace.setBootstrapState(SystemKeyspace.BootstrapState.DECOMMISSIONED);
+            setMode(Mode.DECOMMISSIONED, true);
+            // let op be responsible for killing the process
         }
-        catch (ExecutionException e)
+        catch (Throwable t)
         {
-            logger.error("Error while decommissioning node ", e.getCause());
-            throw new RuntimeException("Error while decommissioning node: " + e.getCause().getMessage());
+            Throwable cause = t.getCause() == null ? t : t.getCause();
+            logger.error("Error while decommissioning node ", cause);
+            SystemKeyspace.setBootstrapState(SystemKeyspace.BootstrapState.DECOMMISSION_FAILED);
```

Comment:
Is it necessary to add Mode.DECOMMISSIONED_FAILED and setMode( Mode.DECOMMISSIONED_FAILED, true) after line 5221?

As talking about the Mode state , Do we need to think again about the state of DECOMMISSIONED FAILED, because every step in the try catch may fail. If there is a failure, are there some corner problems that have not been considered?
Such as when we faild at streaming data , the node statue may show the node is leaving ,and should show users the statues of DECOMMISSIONED FAILED in nodetool status or something else.

Reply:
Answer to your first question: I do not think so.

We have:

    public enum BootstrapState
    {
        NEEDS_BOOTSTRAP,
        COMPLETED,
        IN_PROGRESS,
        DECOMMISSIONED,
        DECOMMISSION_FAILED
    }

But "modes" are:

    public enum Mode { STARTING, NORMAL, JOINING, LEAVING, DECOMMISSIONED, MOVING, DRAINING, DRAINED }

So the question is what consequence it has to introduce "DECOMMISSION_FAILED" state? 

Second question:

I was thinking about this and if you look at it as it is done currently, there is exactly same behavior. It may error out as of now while decommissioning and the problems you described are already there. This patch at least tells us that it failed to decommission what was not the case before.

We may look at this later in separate ticket. I think testing all corner cases is not in the scope of this ticket.

Reply:
> We may look at this later in separate ticket. I think testing all corner cases is not in the scope of this ticket.
Ok

Reply:
Sorry, I have a new question that may need your advice, that is, the initial node joins the ring successfully, and the system table is set with the COMPLETED state. In the old process, if the decommission process is in progress, and the system table has not been updated to need_bootstrap or the final decommissioned ,then the process  decommission failed, the system table is still COMPLETE, but with the new process system table will be updated to decommissioned_failed. 

Although you have updated the judgment of the prepareToJoin function about SystemKeyspace.wasDecommissioned(), there is an isReplacing judgment before prepareToJoin, and the old and new processes may be different in this place. What about decommission failed, and start doing replace node ?
Not sure if my understanding is correct @smiklosovic  @bereng @driftx, can you help to confirm ?
` public boolean isReplacing()
    {
        if (replacing)
            return true;

        if (REPLACE_ADDRESS_FIRST_BOOT.getString() != null && SystemKeyspace.bootstrapComplete())
        {
            logger.info("Replace address on the first boot requested; this node is already bootstrapped");
            return false;
        }

        return DatabaseDescriptor.getReplaceAddress() != null;
    }`

Reply:
What is the current behavior? In trunk, if we decommission and it fails, it stays in `COMPLETED` bootstrap state. 

Then you restart a node and you want to do what? To "replace" a node? Why would you want to replace a node which failed to decommission itself?

I think that is undefined behavior which is currently not covered. I think that it should be like this:

    if (REPLACE_ADDRESS_FIRST_BOOT.getString() != null 
        && (SystemKeyspace.wasDecommissioned()
               || SystemKeyspace.hasDecommissionFailed()
               || SystemKeyspace.bootstrapComplete()))
    {
        return false;


Basically, we should not proceed with replacing procedure if we see that the node was already bootstrapped or it failed to decommission or it was decommissioned.

Reply:
> Then you restart a node and you want to do what? To "replace" a node? Why would you want to replace a node which failed to decommission itself?

I am not going to replace a node after failed to decommission, I am just saying , the newly added STATUS of  decommission_failed can change the old behavior.  

Because , for us, we actually don’t know how users will use it, but what we can guarantee is that if we don’t know how they use it, it will be consistent with the original behavior, or we clearly stated in the document that we will What changes.
But for a managed cassandra service that users do not need to care about the Cassandra's operation and maintenance, of course it is ok .